### PR TITLE
Don't segfault when running atexit before jl_threads_init

### LINKED ISSUE #45726: "Your CPU does not support the CX16 instruction" error segfaults?
https://buildkite.com/julialang/julia-master/builds/12937#018170af-9618-4348-9a71-2ccd3a266d42
```
#15 0x00007f937dd311af in ijl_error (str=str@entry=0x7f937deb6448 "Your CPU does not support the CX16 instruction, which is required by this version of Julia!  This is often due to running inside of a virtualized environment.  Please read https://docs.julialang.org/e"...) at /cache/build/default-amdci4-3/julialang/julia-master/src/rtutils.c:37
```
(this led to a segfault, and hang)

But seems to work okay for me:
```
$ ./julia -C-cx16
ERROR: Your CPU does not support the CX16 instruction, which is required by this version of Julia!  This is often due to running inside of a virtualized environment.  Please read https://docs.julialang.org/en/v1/devdocs/sysimg/ for more.
```

### PR DESCRIPTION
**Pull request by vchuravy**
Co-debugged with @staticfloat on ci-dev call

### CODE REVIEW

File: test/cmdlineargs.jl [hunk 1]
```diff
@@ -249,6 +256,9 @@ let exename = `$(Base.julia_cmd()) --startup-file=no --color=no`
         @test read(`$exename -p2 -t2 -e $code`, String) == "6"
     end
 
+    # Combining --threads and invalid -C should yield a decent error
+    @test !success(`$exename -t 2 -C invalidtarget`)
```
Comment (vchuravy) [inline test/cmdlineargs.jl:260]:
```suggestion
    @test errors_not_signals(`$exename -t 2 -C invalidtarget`)
```

**Comment by vtjnash** [comment]
CI apparently does not like this. It also might be good to stop running unsupported configurations on buildkite. Though we should start to see threads being supported in releases soon.

```
cmdlineargs                               (8) |         failed at 2022-06-20T19:59:01.658
Test Failed at /cache/build/default-amdci4-0/julialang/julia-master/julia-72ed6e5bcd/share/julia/test/cmdlineargs.jl:680
  Expression: errors_not_signals(p)
Test Failed at /cache/build/default-amdci4-0/julialang/julia-master/julia-72ed6e5bcd/share/julia/test/cmdlineargs.jl:682
  Expression: p.exitcode == 1
   Evaluated: -9223372036854775808 == 1
Test Failed at /cache/build/default-amdci4-0/julialang/julia-master/julia-72ed6e5bcd/share/julia/test/cmdlineargs.jl:680
  Expression: errors_not_signals(p)
Test Failed at /cache/build/default-amdci4-0/julialang/julia-master/julia-72ed6e5bcd/share/julia/test/cmdlineargs.jl:682
  Expression: p.exitcode == 1
   Evaluated: -9223372036854775808 == 1
Test Failed at /cache/build/default-amdci4-0/julialang/julia-master/julia-72ed6e5bcd/share/julia/test/cmdlineargs.jl:691
  Expression: errors_not_signals(p)
Test Failed at /cache/build/default-amdci4-0/julialang/julia-master/julia-72ed6e5bcd/share/julia/test/cmdlineargs.jl:693
  Expression: p.exitcode == 1
   Evaluated: -9223372036854775808 == 1
```

```
      From worker 37:	ERROR: option `-e/--eval` is missing an argument
      From worker 37:	ERROR: option `-e/--eval` is missing an argument
      From worker 37:	ERROR: option `-E/--print` is missing an argument
      From worker 37:	ERROR: option `-E/--print` is missing an argument
      From worker 37:	ERROR: option `-L/--load` is missing an argument
      From worker 37:	ERROR: option `-L/--load` is missing an argument
      From worker 37:	fatal: error thrown and no exception handler available.
      From worker 37:	ErrorException("Invalid CPU name "invalidtarget".")
      From worker 37:
      From worker 37:	signal (11): Segmentation fault
      From worker 37:	in expression starting at none:0
      From worker 37:
      From worker 37:	signal (6): Aborted
      From worker 37:	in expression starting at none:0
      From worker 37:	fatal: error thrown and no exception handler available.
      From worker 37:	ErrorException("Invalid CPU name "invalidtarget".")
      From worker 37:
      From worker 37:	signal (11): Segmentation fault
      From worker 37:	in expression starting at none:0
      From worker 37:
      From worker 37:	signal (6): Aborted
      From worker 37:	in expression starting at none:0
```

**Comment by vchuravy** [comment]
How is setting `JULIA_NUM_THREADS` an unsupported configuration?

> CI apparently does not like this.

So yeah this isn't complete :)

**Comment by vtjnash** [comment]
>How is setting JULIA_NUM_THREADS an unsupported configuration?
>>CI apparently does not like this.

Because CI apparently does not like this

File: test/cmdlineargs.jl [hunk 2]
```diff
@@ -666,7 +677,7 @@ let exename = `$(Base.julia_cmd().exec[1]) -t 1`
                 @test !occursin("Segmentation fault", s)
                 @test !occursin("EXCEPTION_ACCESS_VIOLATION", s)
             end
-            @test !success(p)
+            @test errors_not_signals(p)
             @test !Base.process_signaled(p)
```
Comment (giordano) [inline test/cmdlineargs.jl:681]: I think this is made redundant by `errors_not_signals`?  Same at line 692.

**Comment by staticfloat** [comment]
Alright, that latest fix has solved the majority of problems, but i686-linux-gnu still segfaults on `julia -C invalid`.

Digging in with a debugger, we fail at a different point on `i686-linux-gnu`, because [we don't hit the `CX16` failure branch](https://github.com/JuliaLang/julia/blob/3f047f7efb00a555716e2122c71954392685f014/src/processor_x86.cpp#L860-L863), but instead hit [the "Invalid CPU Name" error](https://github.com/JuliaLang/julia/blob/ef7498decee61a84f8df58793ee2c488faaa1074/src/jitlayers.cpp#L823).  This is much later in the bringup process, but still not early enough that we can do things like call `jl_DI_for_fptr`, which is what we try to do when we print backtraces due to calling `jl_errorf()`.

I've now pushed a commit that avoids printing backtraces when `jl_errorf()` is called, in a similar manner to how `jl_error()` avoids backtraces, but I had to extend the guard condition to use other data structures, as it turns out that the existing guard condition was insufficient.  I don't know if this is the proper guard condition (maybe we should create a `jl_can_collect_backtraces` sentinel global value?) but CI will tell us if this works or not.

**Comment by JeffBezanson** [comment]
Maybe we should have jl_init_llvm handle filling in the libjulia-codegen function pointers, so the system just behaves as if codegen is not available before it's initialized?

**Comment by staticfloat** [comment]
Following Jeff's advice on Slack, I have removed the potentially more controversial parts of this PR, manually inlining the "print and exit" logic from `jl_error*` in the cases where we know it is too early to backtrace.  Looking on CI, I got a natural green flush, and have verified that there are no mysterious core files uploaded, so I'm calling this a victory.

**Comment by staticfloat** [comment]
Further discussion about making `jl_error*` safe can be found in this issue: https://github.com/JuliaLang/julia/issues/45847

**Comment by vchuravy** [comment]
Thanks Elliot for getting this over the line!

**Pull request merged**